feat(app): long term metric storage

olevski opened the pull request:
Adds long term metric storage in a s3 bucket.

Handles all metrics in a separate thread.

closes https://github.com/SwissDataScienceCenter/renku-notebooks/issues/1069

Panaetius commented on controller/metrics/s3.py:
```diff
@@ -0,0 +1,145 @@
+        if not self._current_log_start_timestamp:
+            self._current_log_start_timestamp = pytz.UTC.localize(datetime.utcnow())
+        now = pytz.UTC.localize(datetime.utcnow())
+        if (now - self._current_log_start_timestamp).total_seconds() / 3600 > self.period_hours:
```
```suggestion
        if (now - self._current_log_start_timestamp)/timedelta(hours=1) > self.period_hours:
```
is what the datetime docs recommend for units other than seconds

olevski replied:
Now I just use seconds. It is easier that way and it allows me to setup a very short period for testing and development.

Panaetius commented on controller/metrics/s3.py:
```diff
@@ -0,0 +1,145 @@
+        )
+
+
+class BaseMetricLogger(ABC):
```
Did you consider creating a new python `logging` handler for S3 instead? The Info/Debug/Warn model isn't that appropriate for this use case, but it would mean being compatible with many other logging implementations, all the features already available like filtering and I think the basic design of pythons logging is done quite well. At least the shouldRollover/doRollover logic in https://github.com/python/cpython/blob/11e865c74baa45ac9eda69f2f6af05505fc03ce5/Lib/logging/handlers.py#L65 is pretty readable.

I think all that'd be needed is to implement `emit` in a class inheriting from https://github.com/python/cpython/blob/11e865c74baa45ac9eda69f2f6af05505fc03ce5/Lib/logging/__init__.py#L919 and to add the shouldRollover/doRollover logic mentioned above.

But I can see arguments for and against, so just wanted to check if you looked into this. Not something you have to add.

olevski replied:
Yeah I found the docs on `logging` really hard to navigate. But I am pretty sure I figured things out. I have a logging handler that uploads to s3 on rotation. And I have a formatter as well. But yeah as you suggest using logging for this is much nicer.

Panaetius commented on controller/metrics/s3.py:
```diff
@@ -0,0 +1,145 @@
+
+    def publish(self, metric_event: MetricEvent):
+        session_metric_data = SesionMetricData.from_metric_event(metric_event)
+        self.logger.log(session_metric_data.to_json())
+        self.logger.log("\n")
```
```suggestion
        self.logger.log(f"{session_metric_data.to_json()}\n")
```
or
```suggestion
        self.logger.log(session_metric_data.to_json() + "\n)
```
no reason to write twice

olevski replied:
This is not an issue now with the logging handler and formatter.

Panaetius reviewed (changes requested):
We are really lacking class/method docstrings in amalthea, would be nice to have some descriptions on what things do to make them more easily understandable.

olevski commented on controller/metrics/prometheus.py:
```diff
@@ -0,0 +1,227 @@
+            "usage_methods": ["observe"],
+            "type": Summary,
+        },
+    }
```
Yeah enum is def better. Added it.

olevski commented on controller/metrics/prometheus.py:
```diff
@@ -0,0 +1,227 @@
+        sanitized_labels = {}
+        for name, val in labels.items():
+            sanitized_labels[cls.sanitize_label_name(name)] = val
+        return sanitized_labels
```
I added that because they do did not really need `self` of any of the real state of the object. But with some of the other proposed changes now I do need access to the state and there is no more classmethod or staticmethod

olevski commented on controller/metrics/prometheus.py:
```diff
@@ -0,0 +1,227 @@
+        self.manifest_labelnames = manifest_labelnames
+        self._sessions_total_created = PrometheusMetric(
+            "counter",
+            "sessions_total_created",
```
I can convert the names of the metric to an enum. But I think there is no need tbh. They are all defined as class variables at the top and are accessed only through the class variables. 

Should I just make this an enum of metric names and make that a class variable?

olevski replied:
Ok I just made a simple Enum that contains the different metric names. I see what you mean this way there is pretty much zero chance that we mistype a metric name and inadvertently create a new metric. If we were to mistype a metric name then the Enum would error out.

olevski commented on controller/metrics/queue.py:
```diff
@@ -0,0 +1,29 @@
+        while True:
+            metric_event = self.q.get()
+            for handler in self.metric_handlers:
+                handler.publish(metric_event)
```
I added a blanket `except Exception` statement to catch any exception log it as a warning and continue.

olevski commented:
@Panaetius I addressed all the comments. I also added docstrings to all classes and their public methods in the new files for metrics handling.

I just have to fix the tests and that is it.

Panaetius commented on controller/metrics/prometheus.py:
```diff
@@ -157,67 +197,62 @@ def _collect_labels_from_manifest(self, manifest: Dict[str, Any]) -> Dict[str, s
     def _on_start(self, metric_event: MetricEvent):
         manifest_labels = self._collect_labels_from_manifest(metric_event.session)
-        self._sessions_total_created.manipulate("inc", 1, manifest_labels)
+        self._sessions_total_created("inc", 1, manifest_labels)
```
```suggestion
        self._sessions_total_created(PrometheusMetricAction.inc, 1, manifest_labels)
```

Panaetius commented on controller/metrics/prometheus.py:
```diff
@@ -157,67 +197,62 @@ def _collect_labels_from_manifest(self, manifest: Dict[str, Any]) -> Dict[str, s
+        """Publishes (i.e. persists) the proper prometheus metrics
+        depending on the old and new statuses of the jupyterserver."""
         old_status = metric_event.old_status
         new_status = metric_event.session.get("status", {}).get("state")
```
```suggestion
        new_status = metric_event.status
```
?

Panaetius replied:
Not sure about this one, `metric_event.session.get("status", {}).get("state")` might be something else. Ignore if it is

Panaetius reviewed (changes requested):
Thank you, it looks really nice now. Glad that using `logging` turned out to be quite easy :+1: 

Two minor places where I think you forgot to change to the new way of doing things.

Panaetius commented on controller/config_types.py:
```diff
@@ -0,0 +1,66 @@
+
+    def __post_init__(self):
+        if not self.rotation_period_seconds:
+            self.rotation_period_seconds = 86400
```
You can set the default value above instead with `    rotation_period_seconds: Optional[str] = 86400`

Panaetius commented on controller/config_types.py:
```diff
@@ -0,0 +1,66 @@
+        if type(self.extra_labels) is str:
+            self.extra_labels = json.loads(self.extra_labels)
+        if not self.extra_labels:
+            self.extra_labels = []
```
Same here

Panaetius commented on controller/metrics/s3.py:
```diff
@@ -41,9 +32,11 @@ class SesionMetricData:
+
+    def __post_init__(self):
+        if not self.additional_labels:
+            self.additional_labels = {}
```
same here I think

Panaetius commented on controller/metrics/utils.py:
```diff
@@ -20,13 +20,15 @@ def resource_request_from_manifest(manifest: Dict[str, Any]) -> Optional[Resourc
     resources = manifest.get("spec", {}).get("jupyterServer", {}).get("resources", {}).get(
         "requests", {}
     )
+    resources = {**resources}
```
This is essentially a nop? or what does this do?

If it is just to cash to a plain python dict, `resources = dict(resources)` communicates the intent better, imo.

olevski replied:
@Panaetius yeah I essentially wanted a copy of that dictionary. Because in a few lines below I add something to it just for the purposes of this function. But if I do not use a copy then that same dictionary in the manifest gets modified in the same way and breaks things.

Now I just use `copy.deepcopy` for this which is the safest option and properly communicates the intent IMO.

olevski commented:
@Panaetius this is ready now. I went with `dict(**resources)` on making a dictionary copy because `deepcopy` failing to create a copy of resources because in some cases it is a plain dictionary (this works) when the session is first created. But after the session is created and the status changes then this is a dict-like object that I get from querying the k8s api. With `dict` both cases work.
